Add line numbers on all block tokens during parsing (#144)

anderskaplan opened the pull request:
This is a partial solution for #144 which records line numbers on all _block_ tokens. The _span_ tokens are still left unlabeled, but at least it's a starting point, and I think the functionality added here is useful by itself.

Open question: should the `line_number` be a added to `repr_attributes`, so that it gets included in the output from `ASTRenderer`?

pbodnar commented:
@anderskaplan, thank you, I will try to look at this for the next release.

pbodnar commented:
Finally, I have found time for mistletoe once again. :)

@anderskaplan, I must say "good job" 👍 . Basically, I like it, including the clever way of the new unit test.

Probably just 3 more things be resolved:

1. Performance - did you run the benchmark test? I guess there should be little to zero difference to the version without line numbers, right?
2. Optionality - shouldn't we better pass the `line_number` parameter as the last AND optional (with default value of `None`) to the constructors of `TableRow` etc. (even `Table`?)? This way, we could achieve backwards compatibility for those users who might create (some) tokens without actually parsing a markdown text. Yes, there probably won't be many, but what if there are some out there? :)
3. Documentation - probably add some notes on this cool feature as a sub-chapter to https://github.com/miyuchina/mistletoe/blob/v1.2.1/dev-guide.md#understanding-the-ast?

> Open question: should the `line_number` be a added to `repr_attributes`, so that it gets included in the output from `ASTRenderer`?

Not sure here, but probably yes, it could? :)

anderskaplan commented:
Hey @pbodnar, just checking in to say hi. The last weeks have been quite busy with other things but I'll get back to this as soon as I can.

anderskaplan commented:
I have updated the documentation, made the line_number parameter optional, and run the benchmark as requested in the review.

The benchmark results have a fairly large variance (5 %CV) from run to run on my machine. But I'm fairly confident that the performance penalty of the line numbers is negligible, because if I compare the mean benchmark with and without line numbers from five runs of each type, then the relative difference is about 0.1%. I.e., far less than the run-to-run variance.

I also added the `line_number` attribute to `repr_attributes`, so that the line numbers are included in the output from the ASTRenderer.

pbodnar commented on dev-guide.md:
```diff
@@ -86,12 +103,25 @@ Then we will get this JSON output from the AST renderer:
+### Line numbers
+
+mistletoe records the starting line of all block tokens that it encounters during
+parsing and stores it the `line_number` attribute.
```
```diff
-parsing and stores it the `line_number` attribute.
+parsing and stores it as the `line_number` attribute of each token.
```

anderskaplan replied:
Fixed, along with the merge conflicts :)

pbodnar replied:
Perfect, thank you! :)

pbodnar reviewed (changes requested):
@anderskaplan, thank you for the amendments, I've found just one typo. :)

Beside that, it looks like some other conflicting commits were made in the mean time...
